include/wil/cppwinrt.h

```diff
@@ -16,47 +16,70 @@
 #include <unknwn.h>
 #include <hstring.h>
 
-#ifdef CPPWINRT_VERSION
-#error Please include wil/cppwinrt.h before including any C++/WinRT headers
-#endif
+// WIL and C++/WinRT use two different exception types for communicating HRESULT failures. Thus, both libraries need to
+// understand how to translate these exception types into the correct HRESULT values at the ABI boundary. Prior to
+// C++/WinRT "2.0" this was accomplished by injecting the WINRT_EXTERNAL_CATCH_CLAUSE macro - that WIL defines below -
+// into its exception handler (winrt::to_hresult). Starting with C++/WinRT "2.0" this mechanism has shifted to a global
+// function pointer - winrt_to_hresult_handler - that WIL sets automatically when this header is included and
+// 'RESULT_SUPPRESS_STATIC_INITIALIZERS' is not defined.
 
-#ifdef __WIL_RESULTMACROS_INCLUDED
-#error Please include wil/cppwinrt.h before including result_macros.h
-#endif
+/// @cond
+namespace wil::details
+{
+    // Since the C++/WinRT version macro is a string...
+    inline constexpr int major_version_from_string(const char* versionString)
+    {
+        int result = 0;
+        auto str = versionString;
+        while ((*str >= '0') && (*str <= '9'))
+        {
+            result = result * 10 + (*str - '0');
+            ++str;
+        }
 
-#ifdef WINRT_EXTERNAL_CATCH_CLAUSE
-#error C++/WinRT external catch clause already defined outside of WIL
-#endif
+        return result;
+    }
+}
+/// @endcond
 
-// WIL and C++/WinRT use two different exception types for communicating HRESULT failures. Thus, both libraries need to
-// understand how to translate these exception types into the correct HRESULT values at the ABI boundary. C++/WinRT
-// accomplishes this by injecting the WINRT_EXTERNAL_CATCH_CLAUSE macro - that WIL defines below - into its exception
-// handler. WIL accomplishes this by detecting this file's inclusion in result_macros.h and modifies its behavior to
-// account for the C++/WinRT exception type.
+#ifdef CPPWINRT_VERSION
+// Prior to C++/WinRT "2.0" this header needed to be included before 'winrt/base.h' so that our definition of
+// 'WINRT_EXTERNAL_CATCH_CLAUSE' would get picked up in the implementation of 'winrt::to_hresult'. This is no longer
+// problematic, so only emit an error when using a version of C++/WinRT prior to 2.0
+static_assert(::wil::details::major_version_from_string(CPPWINRT_VERSION) >= 2,

```

Comment:
Instead of this string parsing, could cppwinrt produce a `CPPWINRT_VERSION_MAJOR` macro which is a numeric value 2 or similar? I guess that'd require a coordinated change between the repos ... will the version identification mechanism of cppwinrt ever change?

Reply:
Right, it's one more thing we have to version check. This version string has been around for ages so it's rather convenient and avoids introducing yet another macro. And when C++ modules finally comes along we'll be completely macro free, so adding macros is probably not the best strategy in the long run.

Reply:
I'll let Kenny be the authority on which macros are present. But I agree that we should ideally future-proof ourselves for modules, in which case I guess we'll move to something like checking `winrt::version_string` (or integers, but neither are usable with the preprocessor). But by that time, hopefully we can remove all of the back-compat stuff.